### PR TITLE
[Java] Document breaking changes to listCollectionName methods and new authorizedCollections method

### PR DESCRIPTION
**Pull request by rachel-mack**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

JIRA 
 - <https://jira.mongodb.org/browse/DOCSP-35762>
 - Java work: https://jira.mongodb.org/browse/JAVA-4353

Staging
 - [What's New](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-35762/whats-new/#what-s-new-in-5.0)
 - [Breaking Changes](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-35762/upgrade/#version-5.0-breaking-changes)

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

**Review by jordan-smith721 — APPROVED** [review]
LGTM

File: source/upgrade.txt [hunk 1]
```diff
@@ -151,6 +151,11 @@ Version 5.0 Breaking Changes
   responding in the same way as ``Decimal128`` values. In particular,
   ``BsonDecimal128.isNumber()`` now returns ``true``, and
   ``BsonDecimal128.asNumber()`` returns the equivalent ``BsonNumber``.
+.. _java-breaking-changes-v5.0-list-collections:
+
+- This driver version breaks binary compatibility by adding a new
+  ``authorizedCollection`` option to the ``listCollections`` command. This
+  change does not require any changes to source code.
```
Comment (stIncMale) [inline source/upgrade.txt:158]: For a user to know whether his code has to be recompiled due to a change that broke binary compatibility, the user must know whether his code uses the program element that was changed. The current text does not mention that program element (the methods `MongoDatabase.listCollectionNames` were changed in a way that breaks binary compatibility).

File: source/whats-new.txt [hunk 2]
```diff
@@ -67,6 +67,21 @@ The 5.0 driver release introduces the following features:
...
+  
+- The ``listCollections`` command has an ``authorizedCollections`` option. It can be configured using the
+  ``MongoDatabase.listCollectionNames()`` method. This introduces a :ref:`binary
+  breaking change to the listCollections
+  command<java-breaking-changes-v5.0-list-collections>`.
+
+  For more information about ``authorizedCollections``, see the
+  :manual:`listCollections </reference/command/listCollections/>` server manual
+  page. For more information about the ``listCollectionNames()`` method, see the
+  :ref:`Database Collections<java-collection-names>` page.
```
Comment (stIncMale) [inline source/whats-new.txt:83]: These two list items say the same thing, why do they both exist?

File: source/whats-new.txt [hunk 3]
```diff
@@ -67,6 +67,21 @@ The 5.0 driver release introduces the following features:
...
 
+- A new ``authorizedCollections`` option has been added to the
+  ``listCollections`` command. It can be configured using the
+  ``MongoDatabase.listCollectionNames()`` method. This introduces a
+  :ref:`breaking change to the listCollections
+  
+- The ``listCollections`` command has an ``authorizedCollections`` option. It can be configured using the
+  ``MongoDatabase.listCollectionNames()`` method. This introduces a :ref:`binary
+  breaking change to the listCollections
+  command<java-breaking-changes-v5.0-list-collections>`.
```
Comment (stIncMale) [inline source/whats-new.txt:78]: The concept of binary compatibility is not applicable to the `listCollections` MongoDB command, it is in the current context only applicable to the methods `MongoDatabase.listCollectionNames`.

I should have probably wrote a more detailed "Documentation Changes Summary" in https://jira.mongodb.org/browse/JAVA-4353. Apologies, let me fix that mistake here.

The methods `com.mongodb.client.MongoDatabase.listCollectionNames` were returning `MongoIterable<String>`, which did not allow configuring the `authorizedCollections` option of the `listCollections` command. To support this option, the return type of the methods was changed to `com.mongodb.client.ListCollectionNamesIterable`, which broke binary compatibility. Now users may use the method `ListCollectionNamesIterable.authorizedCollections` to configure the `authorizedCollections` option.

A similar change was done to the `listCollectionNames` methods in the following classes/interfaces:

- `com.mongodb.reactivestreams.client.MongoDatabase`
- `org.mongodb.scala.MongoDatabase`
- `com.mongodb.kotlin.client.MongoDatabase`
- `com.mongodb.kotlin.client.coroutine.MongoDatabase`

However, the original and new return types of all the `listCollectionNames` methods are different across all the aforementioned classes/interfaces, which is why I originally formulated the "Documentation Changes Summary" the way I did: it avoids referring to those types by names, while still being correct (yet apparently not clear enough).

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: source/whats-new.txt (hunk 2 above)
Comment (rachel-mack) [reply source/whats-new.txt]: Removed one. I think there was confusion around the difference between the listCollections command and the listCollectionsNames method.

File: source/upgrade.txt [hunk 4]
```diff
@@ -151,6 +152,13 @@ Version 5.0 Breaking Changes
   responding in the same way as ``Decimal128`` values. In particular,
   ``BsonDecimal128.isNumber()`` now returns ``true``, and
   ``BsonDecimal128.asNumber()`` returns the equivalent ``BsonNumber``.
+.. _java-breaking-changes-v5.0-list-collections:
+
+- This driver version adds a new ``authorizedCollection`` option to the
+  ``listCollections`` command. This introduces a breaking binary compatibility
```
Comment (stIncMale) [inline source/upgrade.txt:158]: The driver cannot add any options to server commands, as that is a server's prerogative. I propose the following wording

```suggestion
- This driver version adds support for a new ``authorizedCollection`` option of the
  ``listCollections`` command. This introduces a breaking binary compatibility
```

File: source/whats-new.txt [hunk 5]
```diff
@@ -66,7 +66,27 @@ The 5.0 driver release introduces the following features:
 
 - Changes to the ``ConnectionId`` class. To learn more, see the
   :ref:`Version 5.0 Breaking Changes <java-breaking-changes-v5.0>`.
-
+  
+- Changes to the ``com.mongodb.client.MongoDatabase.listCollectionNames()``
+  methods. The return type is now a
+  ``com.mongodb.client.ListCollectionNamesIterable`` type, while previously it
+  was a ``MongoIterable<String>``. This changes allows the return value to be
```
Comment (stIncMale) [inline source/whats-new.txt:73]: > This changes allows

Should probably be either _"This change allows"_ or _"These changes allow"_.

File: source/whats-new.txt [hunk 6]
```diff
@@ -66,7 +66,27 @@ The 5.0 driver release introduces the following features:
...
+  method and specifying the ``authorizedCollections`` option. Equivalent changes
+  were made to following classes and interfaces:
+
+  - ``com.mongodb.reactivestreams.client.MongoDatabase``
+  - ``org.mongodb.scala.MongoDatabase``
+  - ``com.mongodb.kotlin.client.MongoDatabase``
+  - ``com.mongodb.kotlin.client.coroutine.MongoDatabase``
+
+  These changes introduce a :ref:` binary breaking change to the listCollections
+  command<java-breaking-changes-v5.0-list-collections>`. For more information
```
Comment (stIncMale) [inline source/whats-new.txt:84]: As described in https://github.com/mongodb/docs-java/pull/516#discussion_r1473806252, the binary breaking change was made to the `listCollectionNames` methods, not the `listCollections` command.

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: source/upgrade.txt (hunk 4 above)
Comment (stIncMale) [reply source/upgrade.txt]: The new wording is

> This driver version adds support for a new ``authorizedCollection`` option of the ``listCollectionsNames()`` method.

It's not what I proposed, and it is not correct: the concept of "option" is not applicable to a method.

**Review by stIncMale — CHANGES_REQUESTED** [review]
.

**Review by stIncMale — APPROVED** [review]

**Pull request merged**